src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs

```diff
@@ -0,0 +1,212 @@
+
+                var info = await service.AnalyzeAsync(
+                    document, type, cancellationToken).ConfigureAwait(false);
+                if (info is not null)
+                {
+                    using var _ = ArrayBuilder<CodeAction>.GetInstance(out var codeActions);
+                    await foreach (var implementOptions in GetImplementOptionsAsync(document, info, cancellationToken))
+                    {
+                        var title = GetTitle(implementOptions);
+                        var equivalenceKey = GetEquivalenceKey(info, implementOptions);
+                        codeActions.Add(CodeAction.Create(
+                            title,
+                            cancellationToken => service.ImplementInterfaceAsync(
+                                document, info, implementOptions, generationOptions, cancellationToken),
+                            equivalenceKey));
+                    }
+
+                    if (codeActions.Count > 0)
+                        context.RegisterFixes(codeActions, context.Diagnostics);
+                }
+
+                break;
+            }
+        }
+    }
+
+    private static string GetTitle(ImplementInterfaceOptions options)
+    {
+        if (options.ImplementDisposePattern)
+        {
+            return options.Explicitly
+                ? FeaturesResources.Implement_interface_explicitly_with_Dispose_pattern
+                : FeaturesResources.Implement_interface_with_Dispose_pattern;
+        }
+        else if (options.Explicitly)
+        {
+            return options.OnlyRemaining
+                ? FeaturesResources.Implement_remaining_members_explicitly
+                : FeaturesResources.Implement_all_members_explicitly;
+        }
+        else if (options.Abstractly)
+        {
+            return FeaturesResources.Implement_interface_abstractly;
+        }
+        else if (options.ThroughMember != null)
+        {
+            return string.Format(FeaturesResources.Implement_interface_through_0, options.ThroughMember.Name);
+        }
+        else
+        {
+            return FeaturesResources.Implement_interface;
+        }
+    }
+
+    private static string GetEquivalenceKey(
+        IImplementInterfaceInfo state,
+        ImplementInterfaceOptions options)
+    {
+        var interfaceType = state.InterfaceTypes.First();
+        var typeName = interfaceType.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat);
+        var assemblyName = interfaceType.ContainingAssembly.Name;
+
+        // Legacy part of the equivalence key.  Kept the same to avoid test churn.
+        var codeActionTypeName = options.ImplementDisposePattern
+            ? "Microsoft.CodeAnalysis.ImplementInterface.AbstractImplementInterfaceService+ImplementInterfaceWithDisposePatternCodeAction"
+            : "Microsoft.CodeAnalysis.ImplementInterface.AbstractImplementInterfaceService+ImplementInterfaceCodeAction";
```

Comment:
the tests hardcode these values.  so this allowed us to change the inheritance hierarchy without rewriting *all* the tests.